test/distributed/org/apache/cassandra/distributed/test/repair/ForceRepairTest.java

```diff
@@ -0,0 +1,170 @@
+import org.apache.cassandra.distributed.api.SimpleQueryResult;
+import org.apache.cassandra.distributed.impl.RowUtil;
+import org.apache.cassandra.distributed.shared.AssertUtils;
+import org.apache.cassandra.distributed.shared.ClusterUtils;
+import org.apache.cassandra.distributed.test.TestBaseImpl;
+import org.apache.cassandra.io.sstable.format.SSTableReader;
+import org.apache.cassandra.io.sstable.metadata.StatsMetadata;
+import org.apache.cassandra.schema.Schema;
+import org.apache.cassandra.schema.TableMetadata;
+import org.assertj.core.api.Assertions;
+
+public class ForceRepairTest extends TestBaseImpl
+{
+    /**
+     * Port of python dtest "repair_tests/incremental_repair_test.py::TestIncRepair::test_force" but extends to test
+     * all types of repair.
+     */
+    @Test
+    public void force() throws IOException
+    {
+        force(false);
+    }
+
+    @Test
+    public void forceWithDifference() throws IOException
+    {
+        force(true);
+    }
+
+    private void force(boolean includeDifference) throws IOException
+    {
+        try (Cluster cluster = Cluster.build(3)
+                                      .withConfig(c -> c.set("hinted_handoff_enabled", false)
+                                                        .with(Feature.values()))
+                                      .start())
+        {
+            init(cluster);
+            cluster.schemaChange(withKeyspace("CREATE TABLE %s.tbl (k INT PRIMARY KEY, v INT)"));
+
+            for (int i = 0; i < 10; i++)
+                cluster.coordinator(1).execute(withKeyspace("INSERT INTO %s.tbl (k,v) VALUES (?, ?)"), ConsistencyLevel.ALL, i, i);
+
+            ClusterUtils.stopUnchecked(cluster.get(2));
+
+            // repair should fail because node2 is down
+            IInvokableInstance node1 = cluster.get(1);
+
+            for (String[] args : Arrays.asList(new String[]{ "--full" },
+                                               new String[]{ "--full", "--preview" },
+                                               new String[]{ "--full", "--validate"}, // nothing should be in the repaired set, so shouldn't stream
+                                               new String[]{ "--preview" }, // IR Preview
+                                               new String[]{ "--validate"}, // nothing should be in the repaired set, so shouldn't stream
+                                               new String[0])) // IR
+            {
+                if (includeDifference)
+                    node1.executeInternal(withKeyspace("INSERT INTO %s.tbl (k,v) VALUES (?, ?)"), -1, -1); // each loop should have a different timestamp, causing a new difference
+
+                try
+                {
+                    node1.nodetoolResult(ArrayUtils.addAll(new String[] {"repair", KEYSPACE}, args)).asserts().failure();
+                    node1.nodetoolResult(ArrayUtils.addAll(new String[] {"repair", KEYSPACE, "--force"}, args)).asserts().success();
```

Comment:
General question about `nodetool repair --force`: if a node is down when repair starts, gets ignored, then comes up and handles writes, those writes will not be repaired, even though the SSTables on the repaired nodes will get their last repaired time updated to a time after the node came up and the write was committed. It doesn't look like we check that the set of nodes that were down are still down when finishing the repair and updating the last repaired time. Do you see that posing any problems?

Reply:
> if a node is down when repair starts, gets ignored, then comes up and handles writes, those writes will not be repaired

if the write is CL_ONE then repair will not handle it.  If you do quorum it will be present on another node.

> even though the SSTables on the repaired nodes will get their last repaired time updated to a time after the node came up and the write was committed. 

Repair time and data time have no relation.  Users can control the timestamp and can do things like timestamp=4, if we used this to say "as of timestamp=123577123888 we are fully in-sync" we will not catch those timestamp=4 cases.

Normal repair is blind and checks the full range.  IR repair segregates repaired SSTables from non-repaired ones, so it only needs to check the non-repaired ones (hence why its incremental).

> It doesn't look like we check that the set of nodes that were down are still down when finishing the repair 

too late at this point, we would need to redo the repair again in order to include them